plugins/pulseaudio/PulseAudioDevice.h

```diff
@@ -45,10 +45,15 @@ class AUD_PLUGIN_API PulseAudioDevice : public SoftwareDevice
 	class PulseAudioSynchronizer : public DefaultSynchronizer
 	{
 		PulseAudioDevice* m_device;
+		pa_usec_t m_time_start = 0;
+		double m_seek_pos = 0.0f;
+		double m_timeline_pos = 0.0f;
```

Comment:
A single`pa_usec_t m_start_time = 0` should be enough. When you seek, you can simply compute it by getting the current time and subtracting the seek time (and consider latency as mentioned).

Reply:
I don't think that will work as the time counter from pulseaudio will continue to count up with pause (IIRC because we don't always cork the stream when paused).

Maybe I should rework things more to see if we can cut down on these variables?

Reply:
You're right probably. Let's compare to above: pa_stream_get_time does consider latency, so you don't have to subtract it anymore. However it does not consider the content of the ring buffer, so to be really accurate you would need to update the timing again with the amount of samples stored in the ring buffer at the time of seeking, which again gives you negative positions right after starting playback/negative offsets after seeking and thus the same behavior. Have you actually tried simply using the default synchronizer, i.e., directly returning the playback handle's position and find out which one of theses two is more accurate? Might actually be the same pretty much, since I would assume the latency would be shorter to the extent of insignificance in comparison to the ring buffer size. You could also try the original method with just adapting with the ring buffer size, which would give you almost the same result as your current implementation if you enabled `PA_STREAM_NOT_MONOTONIC` and have a bad delay prediction.

Reply:
> Have you actually tried simply using the default synchronizer, i.e., directly returning the playback handle's position and find out which one of theses two is more accurate?

I have, yes. It doesn't make too much of a difference on low fps counts (24 fps and below). But as you go higher (30, 60, 120 fps) there are jumps/skips in the playback because the position is not updated frequently enough. By using the reported time from `pa_stream_get_time` I get less jumps/skips.
(The pipewire backend is even more accurate in this regard as well.)

> However it does not consider the content of the ring buffer

Hmm, isn't the ring buffer more of a prebuffer?
IE in theory we could fill up the ring buffer with audio data lasting a minute or so because we already know what we are going to playback from the time line.

It should actually reduce the latency because we have already prebuffered the audio data that we are going to send to pulseaudio. So if the ring buffer is already filled, the latency introduced by feeding it to pulseaudio should be minimal.

Or am I getting this wrong?

Of course when seeking or looping, we might need to wait for the ring buffer to clear out and get filled up with new data. However in these cases, we should probably stop counting up the clock until we have filled the ringbuffer and are ready to feed pulseaudio.

For looping sections, I've seen other programs making sure that their audio backend knows about the start/end times so it can pre buffer data from the start when it is approaching the end.

Reply:
I thought a bit about all that. We could actually do this for all backends. By default (if the backend doesn't provide such timing functions), we could simple use the samples played and count from there. That would allow us to remove the handle from the `getPosition` and `seek` methods. We would store the sync callback and call it directly when seek is called (locking the device in the meantime). Of course this can only be as accurate as the buffer size used for the callback, but that should be more precise than the ring buffer which is usually 3x the size.

You don't want to make the ring buffer unnecessarily big. Sure, if it's just the scene playing it would work, but again, the UI could also want to play back sounds in some cases, maybe in plugins and that would then have a 1 minute delay, which is of course unacceptable.

As for the latency that is introduced by the ring buffer, we could simply store the ring buffer read size when play or seek is called and keep the reported position at the start/seek time until we can reduce that to 0. I guess this is what you mean by stop counting up the clock in that case, right?

As for looping sections: this could be done already. In Blender you could put a limiter and a looper on top of the sequencer when you create the playback handle. That would ensure that it loops very precisely. The only thing you have to do then is modulo the reported positions, since even when looping the reported position will increase monotonically. I.e., if you have 250 frames and it reports frame 254, you need to go to frame 254 % 250 = 4.

In that sense, I'm fine with this patch, if that sounds good to you? I would just ask you to do a little clean up and we can merge this. I can implement the changes to the interface and default synchronizer as described at the beginning.